File: bot/warp.py
"""Cloudflare WARP API client."""

from __future__ import annotations

import base64
from dataclasses import dataclass
from datetime import datetime, timezone

import httpx
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey

API_BASE = "https://api.cloudflareclient.com/v0i1909051800"
_TIMEOUT = 30.0


@dataclass(frozen=True)
class WarpAccount:
    """Credentials returned after registering with Cloudflare WARP."""

    private_key: str
    public_key: str
    peer_public_key: str
    client_ipv4: str
    client_ipv6: str


def generate_keys() -> tuple[str, str]:
    """Generate an X25519 key-pair and return ``(private_b64, public_b64)``."""
    private_key = X25519PrivateKey.generate()
    private_bytes = private_key.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    public_bytes = private_key.public_key().public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )
    return (
        base64.b64encode(private_bytes).decode(),
        base64.b64encode(public_bytes).decode(),
    )


async def register_warp() -> WarpAccount:
    """Register a new WARP device and return the account credentials."""
    private_key, public_key = generate_keys()
    now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        # 1. Register a new device
        reg = await client.post(
            f"{API_BASE}/reg",
            headers={
                "Content-Type": "application/json",
                "User-Agent": "okhttp/3.12.1",
            },
            json={
                "install_id": "",
                "tos": now,
                "key": public_key,
                "fcm_token": "",
                "type": "ios",
                "locale": "en_US",
            },
        )
        reg.raise_for_status()
        reg_data = reg.json()

        account_id = reg_data["result"]["id"]
        token = reg_data["result"]["token"]

        # 2. Enable WARP on the account
        cfg = await client.patch(
            f"{API_BASE}/reg/{account_id}",
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {token}",
                "User-Agent": "okhttp/3.12.1",
            },
            json={"warp_enabled": True},
        )
        cfg.raise_for_status()
        cfg_data = cfg.json()

        peer_key = cfg_data["result"]["config"]["peers"][0]["public_key"]
        addresses = cfg_data["result"]["config"]["interface"]["addresses"]

    return WarpAccount(
        private_key=private_key,
        public_key=public_key,
        peer_public_key=peer_key,
        client_ipv4=addresses["v4"],
        client_ipv6=addresses["v6"],
    )

File: web/lib/warp.js
'use strict';

const crypto = require('node:crypto');

const API_BASE = 'https://api.cloudflareclient.com/v0i1909051800';
const TIMEOUT_MS = 30000;

function generateKeys() {
  const { publicKey, privateKey } = crypto.generateKeyPairSync('x25519');
  const privDer = privateKey.export({ type: 'pkcs8', format: 'der' });
  const pubDer = publicKey.export({ type: 'spki', format: 'der' });

  // Raw 32-byte X25519 keys are at the end of the DER encoding
  const privRaw = privDer.subarray(privDer.length - 32);
  const pubRaw = pubDer.subarray(pubDer.length - 32);

  return {
    privateKey: privRaw.toString('base64'),
    publicKey: pubRaw.toString('base64'),
  };
}

async function registerWarp() {
  const keys = generateKeys();
  const now = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');

  // 1. Register a new device
  const regResp = await fetch(`${API_BASE}/reg`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'User-Agent': 'okhttp/3.12.1',
    },
    body: JSON.stringify({
      install_id: '',
      tos: now,
      key: keys.publicKey,
      fcm_token: '',
      type: 'ios',
      locale: 'en_US',
    }),
    signal: AbortSignal.timeout(TIMEOUT_MS),
  });

  if (!regResp.ok) {
    throw new Error(`WARP registration failed: ${regResp.status}`);
  }

  const regData = await regResp.json();
  const accountId = regData.result.id;
  const token = regData.result.token;

  // 2. Enable WARP on the account
  const cfgResp = await fetch(`${API_BASE}/reg/${accountId}`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
      'User-Agent': 'okhttp/3.12.1',
    },
    body: JSON.stringify({ warp_enabled: true }),
    signal: AbortSignal.timeout(TIMEOUT_MS),
  });

  if (!cfgResp.ok) {
    throw new Error(`WARP activation failed: ${cfgResp.status}`);
  }

  const cfgData = await cfgResp.json();
  const peer = cfgData.result.config.peers[0];
  const addresses = cfgData.result.config.interface.addresses;

  return {
    privateKey: keys.privateKey,
    publicKey: keys.publicKey,
    peerPublicKey: peer.public_key,
    clientIpv4: addresses.v4,
    clientIpv6: addresses.v6,
  };
}

module.exports = { generateKeys, registerWarp };
